### PR TITLE
iOS_alignment/retry failed uploads

### LINKED ISSUE #830: [SDL 0191] Retry Failed File Uploads
Proposal: [Retry Failed File Uploads](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0191-retry-failed-file-uploads.md)

> This proposal is to implement retry attempts when a file upload fails.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/564

Steering Committee Decision:

> The Steering Committee voted to accept this proposal with the following revision: the proposal should be updated to include adding this functionality to the Android library as well.
> The proposal was revised to incorporate this revision on August 2, 2018.

### PR DESCRIPTION
**Pull request by JulianKast**
Fixes #830 and #1274 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [X] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [X] I have run the unit tests with this PR
- [X] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).
- [X] I have tested Android, Java SE, and Java EE

#### Unit Tests
Created Unit test in FileManagerTest to simulate SdlFile/SdlArtwork failing to upload, then retrying upload base on FileManagerConfig:
* testFileUploadRetry()
* testArtworkUploadRetry()

Created Unit test in FileManagerTest to simulate a list of files failing and then retrying to upload them:
* testListFilesUploadRetry()

Test for deleting a list of files:
* testDeleteRemoteFilesWithNamesSuccess()
* testDeleteRemoteFilesWithNamesFail()

Fixed some of the old unit test:
testFileUploadSuccess()
testMultipleFileUpload()

More test need to be fixed, going to open issue. I just fixed what was necessary for this feature.

#### Core Tests
* Set FileManagerConfig values artworkRetryCount and fileRetryCount to at-least 2 or higher, (5) would be better and set each value different.(FileManagerConfig is set in sdlManager builder, see Summary below for example)
*  Commented out setting file data in FileManager createPutFile to simulate File Failing
* Created SdlFile with a type of Audio_Wave, log every time file would retry to upload in baseFileManager uploadFile, in the onError of the listener where it recursively calls uploadFile.
* Repeat with an Artwork filetype: BMP, JPEG, PNG.

* Test uploading a list of files the same way except logging where it uploads in BaseFileManager sendMultipleFileOperations, in onFinished() where it recursively calls sendMultipleFileOperations.
* With a list of files test setting one file to an instance of SdlArtwork and the other audio, have each file upload retry count be different and verify that you get both errors back on the developers side.

### Summary
If a file fails to upload, the File will now retry to upload depending on what is set in FileManagerConfig setFileRetryCount() and setFielArtworkRetryCount(). Each value is set to 1 by default and initialized in sdlManager. If you want to set it higher, you can create an instance of FileManagerConfig and set it when you create the SdlManager Builder:

FileManagerConfig fileManagerConfig = new FileManagerConfig();
fileManagerConfig.setArtworkRetryCount(4);
fileManagerConfig.setFileRetryCount(5);

builder.setFileManagerConfig(fileManagerConfig);
sdlManager.start();

##### Enhancements
* Added functionality to retry failed file uploads.

### CLA
- [X] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: base/src/main/java/com/smartdevicelink/managers/file/BaseFileManager.java [hunk 1]
```diff
@@ -240,11 +256,45 @@ public void addCorrelationId(int correlationid) {
...
-				if(listener != null) {
-					if (errors.isEmpty()) {
-						listener.onComplete(null);
+			public void onFinished() { ///tomorrow add logic for delete operations and test!!!
+				if (errors.isEmpty() && customErrors.isEmpty()) {
+					if (listener != null) {
+						listener.onComplete(null); //if first time no errors and listener is not null
+					}
+				} else if (!deletionOperation) {
+					final List<RPCRequest> reRequest = new ArrayList<>();
```
Comment (bilal-alsharifi) [inline base/src/main/java/com/smartdevicelink/managers/file/BaseFileManager.java]: Can we rename this to `requestsToResend`

File: android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/file/FileManagerTests.java [hunk 2]
```diff
@@ -57,6 +60,45 @@ public void tearDown() throws Exception {
...
+		public Void answer(InvocationOnMock invocation) throws Throwable {
+			Object[] args = invocation.getArguments();
+			RPCRequest message = (RPCRequest) args[0];
+			if(message instanceof PutFile){
+				int correlationId = message.getCorrelationID();
+				Result resultCode = Result.REJECTED;
+				PutFileResponse putFileResponse = new PutFileResponse();
+				putFileResponse.setSuccess(false);
+				message.getOnRPCResponseListener().onError(correlationId, resultCode,"Binary data empty");
+			}
```
Comment (bilal-alsharifi) [inline android/sdl_android/src/androidTest/java/com/smartdevicelink/managers/file/FileManagerTests.java]: Can we remove the extra spaces here?

File: base/src/main/java/com/smartdevicelink/managers/file/BaseFileManager.java [hunk 3]
```diff
@@ -212,7 +227,7 @@ public void deleteRemoteFilesWithNames(@NonNull List<String> fileNames, final Mu
 	 * @param requests Non-empty list of PutFile or DeleteFile requests
 	 * @param listener MultipleFileCompletionListener that is called upon conclusion of sending requests
 	 */
-	private void sendMultipleFileOperations(final List<? extends RPCRequest> requests, final MultipleFileCompletionListener listener){
+	private void sendMultipleFileOperations(final List<? extends RPCRequest> requests, final MultipleFileCompletionListener listener, final Map<String, String> customErrors ){
```
[the remark this answers is not in the document]
Comment (JulianKast) [reply base/src/main/java/com/smartdevicelink/managers/file/BaseFileManager.java]: under new logic we only need on hash to handle errors, Im going to get rid of the current hash called errors and rename customErrors to errors.

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/file/FileManager.java [hunk 4]
```diff
@@ -64,14 +64,21 @@
...
-
+	@Deprecated
 	public FileManager(ISdl internalInterface, Context context) {
 
 		// setup
 		super(internalInterface);
 		this.context = new WeakReference<>(context);
 	}
 
+	public FileManager(ISdl internalInterface, Context context,FileManagerConfig fileManagerConfig) {
```
[the remark this answers is not in the document]
Comment (JulianKast) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/file/FileManager.java]: Doen

File: base/src/main/java/com/smartdevicelink/managers/file/BaseFileManager.java [hunk 5]
```diff
@@ -330,7 +429,8 @@ public void uploadFiles(@NonNull List<? extends SdlFile> files, final MultipleFi
 		for(SdlFile file : files){
 			putFileRequests.add(createPutFile(file));
 		}
-		sendMultipleFileOperations(putFileRequests, listener);
+		final Map<String,String>  customErrors = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (JulianKast) [reply base/src/main/java/com/smartdevicelink/managers/file/BaseFileManager.java]: Reformatted all code that I modified to take care of extra spaces

File: base/src/main/java/com/smartdevicelink/managers/file/BaseFileManager.java [hunk 6]
```diff
@@ -274,7 +324,30 @@ public void onResponse(int correlationId, RPCResponse response) {
...
+		};
+
+		internalInterface.sendRequests(requests, onMultipleRequestListener);
+	}
+
+	/**
+	 *  Check to see if PutFile Request can be reUploaded
+	 * @param req PutFile req is a failed file upload
+	 */
+	private boolean checkRequestForReUpload(PutFile req){
```
[the remark this answers is not in the document]
Comment (JulianKast) [reply base/src/main/java/com/smartdevicelink/managers/file/BaseFileManager.java]: Done

File: base/src/main/java/com/smartdevicelink/managers/file/BaseFileManager.java [hunk 7]
```diff
@@ -59,6 +60,7 @@
 import java.io.InputStream;
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.Hashtable;
```
Comment (bilal-alsharifi) [inline base/src/main/java/com/smartdevicelink/managers/file/BaseFileManager.java]: This import is not used

**Pull request merged**